File: dotnet/Net/TCPServer.cs
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

using System;
using System.IO;
using System.Linq;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using System.Net.Security;
using System.Net.NetworkInformation;
using System.Text;
using System.Runtime;
using System.Runtime.Serialization;


namespace Michitai.Lan.Net
{
    /// <summary>
    /// TCP server for managing multiple client connections and message-based communication.
    /// </summary>
    public sealed class TCPServer
    {
        private TcpListener _listner;

        private IPEndPoint _IPEndPoint;

        private int _buffer_size;

        private TCPServerClient[] _clients;

        private bool _closed;

        private readonly object _clients_lock;

        /// <summary>
        /// Event raised when a client connects.
        /// </summary>
        public event Action<string> OnClientConnected;

        /// <summary>
        /// Event raised when a client disconnects.
        /// </summary>
        public event Action<string> OnClientDisconnected;

        /// <summary>
        /// Event raised when a request message is received from a client.
        /// </summary>
        public event Action<IdentifiedMessage> OnRequest;

        /// <summary>
        /// Event raised when the server stops.
        /// </summary>
        public event Action OnStop;

        /// <summary>
        /// Gets whether the server is closed.
        /// </summary>
        public bool IsClosed
    {
        get
        {
            return _closed;
        }

        private set
        {
            _closed = value;
        }
    }

        /// <summary>
        /// Gets the IP endpoint of the server.
        /// </summary>
        public IPEndPoint IpEndPoint => _IPEndPoint;

        /// <summary>
        /// Initializes a new instance of TCPServer with the specified IP address and port.
        /// </summary>
        /// <param name="ip">The IP address to bind to.</param>
        /// <param name="port">The port to listen on.</param>
        /// <param name="buffer_size">The buffer size for network operations.</param>
        public TCPServer(IPAddress ip, int port, int buffer_size = 4096) :
        this(new IPEndPoint(ip, port), buffer_size)
    {
    }

        /// <summary>
        /// Initializes a new instance of TCPServer with the specified IP endpoint.
        /// </summary>
        /// <param name="ip_end_point">The IP endpoint to bind to.</param>
        /// <param name="buffer_size">The buffer size for network operations.</param>
        public TCPServer(IPEndPoint ip_end_point, int buffer_size = 4096)
    {
        _IPEndPoint = ip_end_point;

        _buffer_size = buffer_size;

        _listner = new TcpListener(_IPEndPoint);

        _clients = new TCPServerClient[0];

        IsClosed = false;


        _clients_lock = new object();
    }



        /// <summary>
        /// Starts the TCP server and begins accepting client connections.
        /// </summary>
        public void Start()
    {
        _listner.Start();

        BeginAccept();
    }

        /// <summary>
        /// Stops the TCP server and disconnects all clients.
        /// </summary>
        public void Stop()
    {
        if (IsClosed)
            return;

        IsClosed = true;

        _listner.Stop();

        Socket socket = _listner.Server;

        socket.Close();

        socket.Dispose();

        lock (_clients_lock)
        {
            for (int i = 0; i < _clients.Length; i++)
            {
                _clients[i].Stop();
            }
        }

        OnStop?.Invoke();
    }



        /// <summary>
        /// Sends a response message to a specific client.
        /// </summary>
        /// <param name="identifiedMessage">The identified message to send.</param>
        public void Response(IdentifiedMessage identifiedMessage)
    {
        lock (_clients_lock)
        {
            for (int i = 0; i < _clients.Length; i++)
            {
                if (_clients[i].ID == identifiedMessage.ID)
                {
                    _clients[i].Response(identifiedMessage.Message);

                    return;
                }
            }
        }
    }



        /// <summary>
        /// Disconnects a client by its ID.
        /// </summary>
        /// <param name="id">The client ID to disconnect.</param>
        public void Disconnect(string id)
    {
        lock (_clients_lock)
        {
            for (int i = 0; i < _clients.Length; i++)
            {
                if (_clients[i].ID == id)
                {
                    _clients[i].Stop();
                    break;
                }
            }
        }

        DeleteClient(id);
    }



        /// <summary>
        /// Begins asynchronous acceptance of client connections.
        /// </summary>
        private void BeginAccept()
    {
        try
        {
            _listner.BeginAcceptTcpClient(BeginAcceptTcpClientCallback, null);
        }
        catch
        {
            DebugConsole.LogError("[TCP-Server] Begin Accept Error!");
        }
    }

        /// <summary>
        /// Callback for asynchronous TCP client acceptance.
        /// </summary>
        /// <param name="result">The asynchronous result.</param>
        private void BeginAcceptTcpClientCallback(IAsyncResult result)
    {
        TcpClient client = null;

        try
        {
            client = _listner.EndAcceptTcpClient(result);

            AddClient(new TCPServerClient(client, (im) => OnRequest?.Invoke(im), (id) => Disconnect(id), _buffer_size));
        }
        catch
        {
            DebugConsole.LogError("[TCP-Server] End Accept Error!");
        }

        BeginAccept();
    }

        /// <summary>
        /// Adds a client to the server's client list.
        /// </summary>
        /// <param name="server_client">The server client to add.</param>
        private void AddClient(TCPServerClient server_client)
    {
        lock (_clients_lock)
        {
            TCPServerClient[] clients = _clients;

            _clients = new TCPServerClient[clients.Length + 1];


            for (int i = 0; i < clients.Length; i++)
            {
                _clients[i] = clients[i];
            }

            _clients[clients.Length] = server_client;


            OnClientConnected?.Invoke(server_client.ID);
        }
    }

        /// <summary>
        /// Removes a client from the server's client list.
        /// </summary>
        /// <param name="id">The client ID to remove.</param>
        private void DeleteClient(string id)
    {
        lock (_clients_lock)
        {
            int index = -1;

            for (int i = 0; i < _clients.Length; i++)
            {
                if (_clients[i].ID == id)
                {
                    index = i;

                    break;
                }
            }

            if (index != -1)
            {
                TCPServerClient[] clients = _clients;

                _clients = new TCPServerClient[clients.Length - 1];

                for (int i = 0; i < index; i++)
                {
                    _clients[i] = clients[i];
                }

                for (int i = index; i < _clients.Length; i++)
                {
                    _clients[i] = clients[i + 1];
                }

                OnClientDisconnected?.Invoke(id);
            }
        }
    }
}
}

File: cpp/src/Net/TCPServer.cpp
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

#include <Michitai/Lan/Net/TCPServer.hpp>
#include <Michitai/Lan/Debugging/DebugConsole.hpp>
#include <Michitai/Lan/Guid.hpp>
#include "SocketInternal.hpp"


namespace Michitai::Lan::Net
{
    //===========================================================================================================================================================
    //  TCPServerClient
    //===========================================================================================================================================================

    TCPServerClient::TCPServerClient(std::unique_ptr<Detail::Socket> socket, RequestHandler onRequest,
                                     StopHandler onStop, int bufferSize)
        : _id(NewGuid())
        , _bufferSize(bufferSize)
        , _socket(std::move(socket))
        , _frameBuffer(bufferSize * 2)
        , _onRequest(std::move(onRequest))
        , _onStop(std::move(onStop))
    {
        // Disable Nagle's algorithm — required for 30-60Hz small-message latency.
        _socket->SetNoDelay(true);
        _socket->SetSendBufferSize(65536);
        _socket->SetReceiveBufferSize(65536);
    }

    TCPServerClient::~TCPServerClient()
    {
        Stop();
        JoinThreads();
    }

    int TCPServerClient::PendingWrites() const
    {
        std::lock_guard lock(_sendMutex);
        return static_cast<int>(_sendQueue.size());
    }

    void TCPServerClient::Start()
    {
        Ptr self = shared_from_this();

        _receiveThread = std::thread([self] { self->ReceiveLoop(); });
        _writeThread = std::thread([self] { self->WriteLoop(); });
    }

    void TCPServerClient::Stop()
    {
        {
            std::lock_guard lock(_stateMutex);
            if (_closed.load())
                return;
            _closed.store(true);
        }

        _onRequest = nullptr;

        {
            std::lock_guard lock(_sendMutex);
            while (!_sendQueue.empty())
                _sendQueue.pop();
        }
        _sendCv.notify_all();

        if (_socket)
        {
            _socket->Shutdown();
            _socket->Close();
        }

        if (_onStop)
            _onStop(_id);
    }

    void TCPServerClient::Response(const Message& message)
    {
        if (IsClosed())
            return;

        std::vector<std::uint8_t> frame = Frame::Pack(message.GetMessage());

        {
            std::lock_guard lock(_sendMutex);
            _sendQueue.push(std::move(frame));
        }
        _sendCv.notify_one();
    }

    void TCPServerClient::ReceiveLoop()
    {
        std::vector<std::uint8_t> buffer(static_cast<std::size_t>(_bufferSize));

        while (!IsClosed())
        {
            const int received = _socket->Receive(buffer.data(), static_cast<int>(buffer.size()));

            if (received <= 0)
            {
                Stop();
                return;
            }

            try
            {
                _frameBuffer.Write(buffer.data(), 0, received);

                std::vector<std::uint8_t> payload;
                std::uint8_t flags;

                while (_frameBuffer.TryRead(payload, flags))
                {
                    Message message(Frame::Unpack(payload, flags));

                    try
                    {
                        if (_onRequest)
                            _onRequest(IdentifiedMessage(message, _id));
                    }
                    catch (const std::exception& ex)
                    {
                        Debug::DebugConsole::LogError("[Michitai.Lan][S-REQUEST-HANDLER-ERROR][" +
                                                    std::string(ex.what()) + "]");
                    }
                }
            }
            catch (const std::exception& ex)
            {
                Debug::DebugConsole::LogError("[Michitai.Lan][S-FRAME-ERROR][" +
                                            std::string(typeid(ex).name()) + ":" + ex.what() + "]");
                Stop();
                return;
            }
        }
    }

    void TCPServerClient::WriteLoop()
    {
        std::unique_lock lock(_sendMutex);

        while (true)
        {
            _sendCv.wait(lock, [&] { return _closed.load() || !_sendQueue.empty(); });

            while (!_sendQueue.empty())
            {
                std::vector<std::uint8_t> frame = std::move(_sendQueue.front());
                _sendQueue.pop();

                lock.unlock();

                const bool sent = _socket->SendAll(frame.data(), static_cast<int>(frame.size()));

                lock.lock();

                if (!sent || _closed.load())
                {
                    lock.unlock();
                    Stop();
                    return;
                }
            }

            if (_closed.load())
                return;
        }
    }

    void TCPServerClient::JoinThreads()
    {
        const std::thread::id current = std::this_thread::get_id();

        if (_receiveThread.joinable())
        {
            if (_receiveThread.get_id() == current)
                _receiveThread.detach();
            else
                _receiveThread.join();
        }

        if (_writeThread.joinable())
        {
            if (_writeThread.get_id() == current)
                _writeThread.detach();
            else
                _writeThread.join();
        }
    }


    //===========================================================================================================================================================
    //  TCPServer
    //===========================================================================================================================================================

    TCPServer::TCPServer(const IPEndPoint& endPoint, int bufferSize)
        : _endPoint(endPoint)
        , _bufferSize(bufferSize)
    {
        _listener = std::make_unique<Detail::Socket>(Detail::Socket::TcpV4());
        _listener->SetReuseAddress(true);

        if (!_listener->Bind(_endPoint))
            Detail::ThrowSocketError("bind");

        _listener->Listen();
    }

    TCPServer::~TCPServer()
    {
        Stop();
        JoinThreads();
    }

    int TCPServer::PendingWrites() const
    {
        std::lock_guard lock(_clientsMutex);

        int sum = 0;
        for (const TCPServerClient::Ptr& client : _clients)
        {
            sum += client->PendingWrites();
        }
        return sum;
    }

    void TCPServer::Start()
    {
        // Reflect the actual bound endpoint (matters when port 0 was requested).
        try
        {
            if (std::optional<IPEndPoint> bound = _listener->LocalEndPoint())
                _endPoint = *bound;
        }
        catch (...)
        {
        }

        Ptr self = shared_from_this();
        _acceptThread = std::thread([self] { self->AcceptLoop(); });
    }

    void TCPServer::Stop()
    {
        if (_closed.exchange(true))
            return;

        if (_listener)
        {
            _listener->Shutdown();
            _listener->Close();
        }

        {
            std::lock_guard lock(_clientsMutex);
            for (const TCPServerClient::Ptr& client : _clients)
            {
                client->Stop();
            }
        }

        OnStop.Invoke();
    }

    void TCPServer::Response(const IdentifiedMessage& identifiedMessage)
    {
        {
            std::lock_guard lock(_clientsMutex);
            for (const TCPServerClient::Ptr& client : _clients)
            {
                if (client->ID() == identifiedMessage.ID)
                {
                    client->Response(identifiedMessage.Message);
                    return;
                }
            }
        }

        Debug::DebugConsole::LogWarning("[TCP-Server] Response dropped: no client with ID '" +
                                      identifiedMessage.ID + "'.");
    }

    void TCPServer::Disconnect(const std::string& id)
    {
        {
            std::lock_guard lock(_clientsMutex);
            for (const TCPServerClient::Ptr& client : _clients)
            {
                if (client->ID() == id)
                {
                    client->Stop();
                    break;
                }
            }
        }

        DeleteClient(id);
    }

    void TCPServer::AcceptLoop()
    {
        while (!IsClosed())
        {
            IPEndPoint remote;
            Detail::Socket socket = _listener->Accept(&remote);

            if (!socket.Valid())
            {
                if (IsClosed())
                    return;

                Debug::DebugConsole::LogError("[TCP-Server] Accept Error!");
                continue;
            }

            try
            {
                std::weak_ptr<TCPServer> weak = shared_from_this();

                TCPServerClient::Ptr client = TCPServerClient::Ptr(new TCPServerClient(
                    std::make_unique<Detail::Socket>(std::move(socket)),
                    [weak](const IdentifiedMessage& message)
                    {
                        if (Ptr server = weak.lock())
                            server->OnRequest.Invoke(message);
                    },
                    [weak](const std::string& id)
                    {
                        if (Ptr server = weak.lock())
                            server->Disconnect(id);
                    },
                    _bufferSize));

                // Register BEFORE starting reads — a read completing during
                // construction could raise OnRequest before the client is routable,
                // and its response would be silently dropped.
                AddClient(client);

                client->Start();
            }
            catch (...)
            {
                Debug::DebugConsole::LogError("[TCP-Server] Accept Error!");
            }
        }
    }

    void TCPServer::AddClient(const TCPServerClient::Ptr& client)
    {
        {
            std::lock_guard lock(_clientsMutex);
            _clients.push_back(client);
        }

        OnClientConnected.Invoke(client->ID());
    }

    void TCPServer::DeleteClient(const std::string& id)
    {
        bool removed = false;

        {
            std::lock_guard lock(_clientsMutex);
            for (auto it = _clients.begin(); it != _clients.end(); ++it)
            {
                if ((*it)->ID() == id)
                {
                    _clients.erase(it);
                    removed = true;
                    break;
                }
            }
        }

        if (removed)
            OnClientDisconnected.Invoke(id);
    }

    void TCPServer::JoinThreads()
    {
        if (_acceptThread.joinable())
        {
            if (_acceptThread.get_id() == std::this_thread::get_id())
                _acceptThread.detach();
            else
                _acceptThread.join();
        }
    }
}
